src/workerd/jsg/async-context.c++

```diff
@@ -0,0 +1,254 @@
+  if (!check(promise->HasPrivate(context, handle))) {
+    // Otherwise, we have to create an opaque wrapper holding a ref to the current frame
+    // because we do not have the option of using an internal field with promises.
+    auto frame = kj::addRef(AsyncContextFrame::current(js));
+    KJ_ASSERT(check(promise->SetPrivate(context, handle, wrapOpaque(context, kj::mv(frame)))));
+  }
+  return promise;
+}
+
+kj::Maybe<Value&> AsyncContextFrame::get(StorageKey& key) {
+  KJ_ASSERT(!key.isDead());
+  storage.eraseAll([](const auto& entry) { return entry.key->isDead(); });
+  return storage.find(key).map([](auto& entry) -> Value& { return entry.value; });
+}
+
+AsyncContextFrame::Scope::Scope(Lock& js, AsyncContextFrame& resource)
+    : Scope(js.v8Isolate, resource) {}
+
+AsyncContextFrame::Scope::Scope(v8::Isolate* isolate, AsyncContextFrame& resource)
+    : isolate(IsolateBase::from(isolate)) {
+  this->isolate.pushAsyncFrame(resource);
+}
+
+AsyncContextFrame::Scope::~Scope() noexcept(false) {
+  isolate.popAsyncFrame();
+}
+
+AsyncContextFrame::StorageScope::StorageScope(
+    Lock& js,
+    StorageKey& key,
+    Value store)
+    : frame(AsyncContextFrame::create(js, nullptr, StorageEntry {
+        .key = kj::addRef(key),
+        .value = kj::mv(store)
+      })),
+      scope(js, *frame) {}
+
+void IsolateBase::pushAsyncFrame(AsyncContextFrame& next) {
+  asyncFrameStack.push_front(&next);
+}
+
+void IsolateBase::popAsyncFrame() {
+  asyncFrameStack.pop_front();
+  KJ_DASSERT(!asyncFrameStack.empty(), "the async context frame stack was corrupted");
+}
+
+void IsolateBase::promiseHook(v8::PromiseHookType type,
+                              v8::Local<v8::Promise> promise,
+                              v8::Local<v8::Value> parent) {
+  auto isolate = promise->GetIsolate();
+
+  // V8 will call the promise hook even while execution is terminating. In that
+  // case we don't want to do anything here.
+  if (isolate->IsExecutionTerminating() || isolate->IsDead()) {
+    return;
+  }
+
+  auto& js = Lock::from(isolate);
+
+  switch (type) {
+    case v8::PromiseHookType::kInit: {
+      // The kInit event is triggered by v8 when a deferred Promise is created. This
+      // includes all calls to `new Promise(...)`, `then()`, `catch()`, `finally()`,
+      // uses of `await ...`, `Promise.all()`, etc.
+      // Whenever a Promise is created, we associate it with the current AsyncContextFrame.
+      KJ_DASSERT(AsyncContextFrame::tryUnwrap(js, promise) == nullptr);
+      AsyncContextFrame::wrap(js, promise);
+      break;
+    }
+    case v8::PromiseHookType::kBefore: {
+      // The kBefore event is triggered immediately before a Promise continuation.
+      // We use it here to enter the AsyncContextFrame that was associated with the
+      // promise when it was created.
+      auto& frame = KJ_ASSERT_NONNULL(AsyncContextFrame::tryUnwrap(js, promise),
```

Comment:
Arguably, it could make sense to remove the async context from the promise once we've unwrapped it, since the continuation is only called once, and that's the only time we need it. This might allow contexts to be GC'd a little bit faster in cases where the promise object is kept around for a while after it resolves.

Reply:
Only if the promise is not rejected and the rejection is unhandled because the context needs to be propagated also to the unhandled rejection handler. If we do detach, we'd only want to do so in the `kAfter` event and only if the promise state is `kFulfilled`.